core/src/main/java/com/graphhopper/reader/osm/WaySegmentParser.java

```diff
@@ -224,15 +223,22 @@ public void handleNode(ReaderNode node) {
 
             acceptedNodes++;
 
-            // we keep node tags for barrier nodes
+            // set a special tag for nodes we want to split
             if (splitNodeFilter.test(node)) {
                 if (nodeType == JUNCTION_NODE) {
                     LOGGER.debug("OSM node {} at {},{} is a barrier node at a junction. The barrier will be ignored",
                             node.getId(), Helper.round(node.getLat(), 7), Helper.round(node.getLon(), 7));
                     ignoredSplitNodes++;
                 } else
-                    nodeData.setTags(node);
+                    node.setTag("gh:split_node", true);
```

Comment:
> I thought about replacing the array+Map with all the tags with an adapted EdgeKVStorage for a more compact storage and see if with this storage the speed is better. And this storage does not yet provide a way to remove a single tag. But shouldn't be hard to implement this...

I think you can just ignore the issue of multiple barrier edges per node for the performance analysis. One more idea for the performance check: Maybe try to restrict the set of node tags we keep to see if the speed depends on the number of nodes we store tags for (number of cases where we create a hash map instead of (re-)using emptyMap(), the number of tags per node. If not there must be something else?

>  I thought we need this gh:split_node marker to avoid creating multiple barrier edges and this can only happen (please correct me if I'm wrong I just try to understand this in more detail :) ) if the node is a junction node?

No, if we keep all the tags we won't need it and we can still make sure we ignore barrier nodes at junctions. And we do not need to traverse the graph or anything to prevent duplicate barrier edges at a single node.

Reply:
My current best guess for the slow down is that there are some tags that occur for almost every node so `LongIntMap nodeTagIndicesByOsmNodeIds;` grows very large and the lookup becomes slow. The simple fix would be figuring out what tags these are and ignoring them, or only allowing specific ones. I would assume that the tags we are actually interested (like crossing etc.) only occur on a small fraction of nodes.

Reply:
Looking at the list of tags in my initial post here, it looks like an accept-list will be better than an ignore-list as there seem to be lots of tags that occur frequently and only a few of them seem useful at all. Also the list might strongly depend on the use-case so it seems like it would be helpful to be able to configure this. 

Did you try if there is a big slow down if we only accept `highway`, `crossing`, `barrier` and `ford` for example?

Reply:
> But this isn't yet implemented in this branch?

No, not yet, but I will try to do it.

Reply:
We don't need the remove call when we keep the information whether a node was split already in a separate set like I did here: 0e4f70fdc81d42f6340ce8de7f31f118f88db526